beater/java_attacher/java_attacher.go

```diff
@@ -24,58 +24,419 @@ import (
-var javaAttacher = filepath.FromSlash("./java-attacher.jar")
+// discoverAllRunningJavaProcesses returns a map of PIDs to running Java processes,
+// by looking for processes with "java" in the command.
+func (j *JavaAttacher) discoverAllRunningJavaProcesses(ctx context.Context) (map[int]*jvmDetails, error) {
+	processes, err := sysinfo.Processes()
+	if err != nil {
+		return nil, err
+	}
+	jvms := make(map[int]*jvmDetails)
+	for _, proc := range processes {
+		pid := proc.PID()
+		info, err := proc.Info()
+		if err != nil {
+			// We intentionally do not log this error, because it is expected that when
+			// apm-server runs as an unprivileged process, it cannot obtain information
+			// about every process on the system.
+			continue
+		}
+		if !strings.Contains(strings.ToLower(info.Exe), "java") {
+			continue
+		}
+		userInfo, err := proc.User()
+		if err != nil {
+			j.logger.Warnf("failed to obtain user for process %d, ignoring", pid)
+			continue
+		}
+		uid := userInfo.EUID
+		if uid == "" {
+			uid = userInfo.UID
+		}
+		gid := userInfo.EGID
+		if gid == "" {
+			gid = userInfo.GID
+		}
 
-func (j JavaAttacher) Run(ctx context.Context) error {
-	cmd := j.build(ctx)
-	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+		var username string
+		if user, err := user.LookupId(uid); err != nil {
+			j.logger.Warnf("failed to obtain username for user %s (process %d), using numerical ID", uid, pid)
+			username = userInfo.EUID
+		} else {
+			username = user.Username
+		}
+
+		jvms[pid] = &jvmDetails{
+			user:        username,
+			uid:         uid,
+			gid:         gid,
+			pid:         pid,
+			startTime:   info.StartTime,
+			command:     info.Exe,
+			version:     "unknown", // filled in later
+			cmdLineArgs: strings.Join(info.Args, " "),
+		}
+	}
+	return jvms, nil
+}
+
+// foreachJVM calls f for each JVM in jvms concurrently,
+// returning when one of the following is true:
+//   1. The function completes for each JVM
+//   2. The timeout is reached.
+//   3.	The context is cancelled.
+//
+// If removeOnError is true and f returns an error, that
+// JVM will be removed from the map.
+func (j *JavaAttacher) foreachJVM(
+	ctx context.Context,
+	jvms map[int]*jvmDetails,
+	f func(ctx context.Context, jvm *jvmDetails) error,
+	removeOnError bool,
+	timeout time.Duration,
+) error {
+	if len(jvms) == 0 {
+		return nil
+	}
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	// mutex for synchronisation removal on error. This being
+	// local assumes jvms is not accessed concurrently by any
+	// other goroutines.
+	var mu sync.Mutex
+
+	var g errgroup.Group
+	for pid, jvm := range jvms {
+		pid, jvm := pid, jvm // copy for closure
+		g.Go(func() error {
+			err := f(ctx, jvm)
+			if err != nil {
+				j.logger.Error(err)
+				if removeOnError {
+					mu.Lock()
+					delete(jvms, pid)
+					mu.Unlock()
+				}
+			}
+			return err
+		})
+	}
+	return g.Wait()
```

Comment:
@axw One of the things that were lost with this better code is the active listening on the context, thus aborting quickly if it is closed. Here we wait up to the timeout, which may be 30 seconds for attachment. 
The thing is that `runAttacherCommand` can potentially block for long on `scanner.Scan()`, which may delay abort. Nevertheless, IIUC, the attach command itself would be killed once the context is closed, which means that `scanner.Scan()` should get `EOF` quite quickly. Is that correct?

Reply:
> Nevertheless, IIUC, the attach command itself would be killed once the context is closed, which means that scanner.Scan() should get EOF quite quickly. Is that correct?

That's right. The code now is dependant on the `f` in `f(ctx, jvm)` doing the right thing, and returning quickly once ctx is cancelled. That's either if the passed-in ctx was cancelled, or if the timeout triggers; whichever comes first. `CommandContext` will honour the ctx's cancellation.